openmc/material.py

```diff
@@ -0,0 +1,1109 @@
+
+        """
+
+        if self._macroscopic is not None:
+            msg = 'Unable to add an S(a,b) table to Material ID="{}" as a ' \
+                  'macroscopic data-set has already been added'.format(self._id)
+            raise ValueError(msg)
+
+        if not isinstance(name, str):
+            msg = 'Unable to add an S(a,b) table to Material ID="{}" with a ' \
+                        'non-string table name "{}"'.format(self._id, name)
+            raise ValueError(msg)
+
+        cv.check_type('S(a,b) fraction', fraction, Real)
+        cv.check_greater_than('S(a,b) fraction', fraction, 0.0, True)
+        cv.check_less_than('S(a,b) fraction', fraction, 1.0, True)
+
+        new_name = openmc.data.get_thermal_name(name)
+        if new_name != name:
+            msg = 'OpenMC S(a,b) tables follow the GND naming convention. ' \
+                  'Table "{}" is being renamed as "{}".'.format(name, new_name)
+            warnings.warn(msg)
+
+        self._sab.append((new_name, fraction))
+
+    def make_isotropic_in_lab(self):
+        self.isotropic = [x[0] for x in self._nuclides]
+
+    def get_nuclides(self):
+        """Returns all nuclides in the material
+
+        Returns
+        -------
+        nuclides : list of str
+            List of nuclide names
+
+        """
+        return [x[0] for x in self._nuclides]
+
+    def get_nuclide_densities(self):
+        """Returns all nuclides in the material and their densities
+
+        Returns
+        -------
+        nuclides : dict
+            Dictionary whose keys are nuclide names and values are 3-tuples of
+            (nuclide, density percent, density percent type)
+
+        """
+
+        nuclides = OrderedDict()
+
+        for nuclide, density, density_type in self._nuclides:
+            nuclides[nuclide] = (nuclide, density, density_type)
+
+        return nuclides
+
+    def get_nuclide_atom_densities(self):
+        """Returns all nuclides in the material and their atomic densities in
+        units of atom/b-cm
+
+        Returns
+        -------
+        nuclides : dict
+            Dictionary whose keys are nuclide names and values are tuples of
+            (nuclide, density in atom/b-cm)
+
+        """
+
+        # Expand elements in to nuclides
+        nuclides = self.get_nuclide_densities()
+
+        sum_density = False
+        if self.density_units == 'sum':
+            sum_density = True
+            density = 0.
+        elif self.density_units == 'macro':
+            density = self.density
+        elif self.density_units == 'g/cc' or self.density_units == 'g/cm3':
+            density = -self.density
+        elif self.density_units == 'kg/m3':
+            density = -0.001 * self.density
+        elif self.density_units == 'atom/b-cm':
+            density = self.density
+        elif self.density_units == 'atom/cm3' or self.density_units == 'atom/cc':
+            density = 1.E-24 * self.density
```

Comment:
this method is kinda nontrivial - consider a helper-method that returns a `(density, sum_density)` tuple to keep this logic tight

```
def get_nuclide_atom_densities(self):
  (density, sum_density) = self._unit_based_density()
```

naming this is hard but that's kinda the point :smiley: